src/helpers.rs

```diff
@@ -827,6 +881,123 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                try_carg_from_prim_and_return!(hir::PrimTy::Int(IntTy::I16), k.to_i16(), CArg::Int16, arg_type);
+                try_carg_from_prim_and_return!(hir::PrimTy::Int(IntTy::I32), k.to_i32(), CArg::Int32, arg_type);
+                try_carg_from_prim_and_return!(hir::PrimTy::Int(IntTy::I64), k.to_i64(), CArg::Int64, arg_type);
+                // the uints
+                try_carg_from_prim_and_return!(hir::PrimTy::Uint(UintTy::U8), k.to_u8(), CArg::UInt8, arg_type);
+                try_carg_from_prim_and_return!(hir::PrimTy::Uint(UintTy::U16), k.to_u16(), CArg::UInt16, arg_type);
+                try_carg_from_prim_and_return!(hir::PrimTy::Uint(UintTy::U32), k.to_u32(), CArg::UInt32, arg_type);
+                try_carg_from_prim_and_return!(hir::PrimTy::Uint(UintTy::U64), k.to_u64(), CArg::UInt64, arg_type);
+                // if no primitives were returned then we have an unsupported type
+                throw_unsup_format!(
+                    "Unsupported scalar argument type to external C function: {:?}",
+                    scalar_res
+                );
+            }
+            _ => {
+                throw_unsup_format!(
+                    "Unsupported argument type to external C function: {:?}",
+                    scalar_res
+                );
+            }
+        }
+    }
+
+    // call external C function and
+    // store output, depending on return type in the function signature
+    fn call_external_c_and_store_return<'a>(
+        &mut self,
+        external_fct_defn: ExternalCFuncDeclRep<'tcx>,
+        dest: &PlaceTy<'tcx, Tag>,
+        ptr: CodePtr,
+        libffi_args: Vec<libffi::high::Arg<'a>>,
+    ) -> InterpResult<'tcx, ()> {
+        let this = self.eval_context_mut();
+        unsafe {
+            // ints
+            call_and_store_int_return!(this, hir::PrimTy::Int(IntTy::I8), i8, dest, libffi_args, ptr, external_fct_defn.output_type);
+            call_and_store_int_return!(this, hir::PrimTy::Int(IntTy::I16), i16, dest, libffi_args, ptr, external_fct_defn.output_type);
+            call_and_store_int_return!(this, hir::PrimTy::Int(IntTy::I32), i32, dest, libffi_args, ptr, external_fct_defn.output_type);
+            call_and_store_int_return!(this, hir::PrimTy::Int(IntTy::I64), i64, dest, libffi_args, ptr, external_fct_defn.output_type);
+            // uints
+            call_and_store_int_return!(this, hir::PrimTy::Uint(UintTy::U8), u8, dest, libffi_args, ptr, external_fct_defn.output_type);
+            call_and_store_int_return!(this, hir::PrimTy::Uint(UintTy::U16), u16, dest, libffi_args, ptr, external_fct_defn.output_type);
+            call_and_store_int_return!(this, hir::PrimTy::Uint(UintTy::U32), u32, dest, libffi_args, ptr, external_fct_defn.output_type);
+            call_and_store_int_return!(this, hir::PrimTy::Uint(UintTy::U64), u64, dest, libffi_args, ptr, external_fct_defn.output_type);
```

Comment:
Instead of trying for each type, inline the macro here, match the `PrimTy` once and dispatch from there.